### PR TITLE
[GLCC]Part-2: @to_static support for PyLayer

### PR DESCRIPTION
**Pull request by MarioLulab**
### PR types

New features

### PR changes

Others

#### Original requirements

It is expected that the PyLayer mechanism in the dynamic graph of the Paddle can interoperate with the @to_staitc of the Paddle's dynamic to static, supporting the custom layer of the PyLayer in the model to be perceived by the @to_static and correctly generating the static graph Program, and supporting the training of the dynamic to static and the exporting for inference. 

So our main objective is to **Enable PyLayer to Support @to_static**. In this PR, we support `@to_static` to percept and translate the logic of forward function and backward function from `PyLayer` and also make effort to percept and process `PyLayerContext` in dy2st module.

#### Related issues

【Feature Request】 PyLayer 功能支持动转静 @to_static 🚀: https://github.com/PaddlePaddle/Paddle/issues/54120

#### Related PR

[GLCC]Part-1: Add pylayer op to Support @to_static: https://github.com/PaddlePaddle/Paddle/pull/56108

#### Project Records

https://github.com/MarioLulab/GLCC2023-Paddle-Record

#### Main works

1. Enable `@to_static` to sense and handle the computational logic in the `forward()` and `backward()` functions in dygraph PyLayer, by modifying the function `convert_load` and returning `StaticPyLayer`. `StaticPyLayer` can complete dy2st code translation, and `StaticPyLayer.apply` can create forward block and backward block.
2. We subtitude `PyLayerContext` with `StaticPyLayerContext` to percept the logic of `PyLayerContext`. And we use the `STEP_SCOPES` and `skip_gc_vars` to solve the problem of pass intermediate variables between forward and backward. We modify the code of garbage collection to support keeping neccessary Variables in step_scope after finishing executing forward graph.
3. Add some testcases of `@to_static` support for `PyLayer` and `static_pylayer` API  
4. Complete some TODO work and polish the code mentioned by the comments of https://github.com/PaddlePaddle/Paddle/pull/56108

#### Future works

1. Refine the `prune.cc` to support pylayer op for `jit.save()` for inference.
2. `static_pylayer` API support not only positional argument but also keyword argument.

3. `static_pylayer` API support not only `paddle.Tensor` but also non-Tensor objects will be included in the argument list.
4. `static_pylayer` API support `stop_gradient = False` for some inputs and `stop_gradient = True` for the other inputs. 
5. `StaticPyLayerContext` support more features (e.g. `mark_not_inplace`) which align with dygraph PyLayer

### CODE REVIEW

File: paddle/fluid/framework/executor.cc [hunk 1]
```diff
@@ -52,6 +53,8 @@ void ExecutorPrepareContext::PrepareUnusedVars(
   if (prog_.Size() > 1) {
     operators::PrepareSafeEagerDeletionOnConditionalOpAndConditionalGradOp(
         prog_, block_id_, ops_);
+    operators::PrepareSafeEagerDeletionOnPyLayerOpAndPyLayerGradOp(
```
Comment (Aurelius84) [inline paddle/fluid/framework/executor.cc]: executor 已经不再使用了，这里可能不需要再适配了？

File: paddle/fluid/framework/ir/graph_helper.cc [hunk 2]
```diff
@@ -707,7 +707,8 @@ static void GetGraphOpDesc(const std::vector<Node *> &nodes,
       if (n->Name() == "while" || n->Name() == "while_grad" ||
           n->Name() == "conditional_block" ||
           n->Name() == "conditional_block_grad" || n->Name() == "recurrent" ||
-          n->Name() == "recurrent_grad") {
+          n->Name() == "recurrent_grad" || n->Name() == "pylayer" ||
+          n->Name() == "pylayer_grad") {
```
Comment (Aurelius84) [inline paddle/fluid/framework/ir/graph_helper.cc]: 这个可以在函数里定义一个 unordered_set<std::string>，然后用set的判断取代 || 。因为这里太长了

File: paddle/fluid/framework/ir/memory_optimize_pass/pylayer_op_eager_deletion_pass.cc [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        target_ops[compute_op->GetScopeIdx()].first.emplace_back(
+            compute_op->GetOp());
+      } else if (compute_op->Name() == "pylayer_grad") {
+        target_ops[compute_op->GetScopeIdx()].second.emplace_back(
+            compute_op->GetOp());
+      }
+    }
+
+    // NOTE(Aurelius84): In case of @to_static, after we finish executing
+    // forward graph, some necessaray variable in step_scope of controlflow_op
```
Comment (Aurelius84) [inline paddle/fluid/framework/ir/memory_optimize_pass/pylayer_op_eager_deletion_pass.cc]:
```suggestion
    // forward graph, some necessaray variable in step_scope of pylayer_op
```

File: paddle/fluid/operators/controlflow/pylayer_op_helper.cc [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+namespace paddle {
+namespace framework {
+class ProgramDesc;
+}  // namespace framework
+}  // namespace paddle
+
+namespace paddle {
+namespace operators {
+
+static bool IsMatchedPyLayerOpAndPyLayerGradGradOp(const OpVariant &fwd_op,
```
Comment (Aurelius84) [inline paddle/fluid/operators/controlflow/pylayer_op_helper.cc]: GradGradOp，是多了一个Grad么

File: python/paddle/jit/dy2static/convert_operators.py [hunk 5]
```diff
@@ -47,6 +49,9 @@ def convert_load(x):
         """
         return _convert_into_variable(x)
 
+    if in_declarative_mode() and isinstance(x, PyLayerMeta):
```
Comment (Aurelius84) [inline python/paddle/jit/dy2static/convert_operators.py]: 这里的in_declarative_mode()是不是可以单独一个if 判断，下面再分三个判断？因为我看都是 in_declarative_mode() and XXX

File: python/paddle/jit/dy2static/py_layer.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+from paddle.fluid.framework import Variable
+from paddle.static.nn import static_pylayer
+
+from .program_translator import convert_to_static, unwrap_decorators
+
+
+def is_pylayer_func(func):
+    """predict whether a function is from PyLayer."""
+    func_self = getattr(func, '__self__', None)
+    if func_self and isinstance(func_self, PyLayerMeta):
```
Comment (Aurelius84) [inline python/paddle/jit/dy2static/py_layer.py]:
```suggestion
    return  func_self and isinstance(func_self, PyLayerMeta):
```
直接这样就可以？

File: python/paddle/static/nn/static_pylayer.py [hunk 7]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-
+import paddle
```
Comment (Aurelius84) [inline python/paddle/static/nn/static_pylayer.py:15]: 这个import 不是必须的？我看后面并没有用到 paddle.xxx

File: test/dygraph_to_static/test_pylayer.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+
+"""Tests for PyLayer of Dynamic-to-Static.
+Only test simple cases here."""
+
+import unittest
+
+import numpy as np
+
+import paddle
+from paddle import fluid
```
Comment (Aurelius84) [inline test/dygraph_to_static/test_pylayer.py]: 不建议依赖fluid，这个已经退场了

File: test/dygraph_to_static/test_pylayer.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+        z = paddle.tanh(data)
+        z = cus_tanh_1.apply(z)
+        return z
+
+
+class TestPyLayerBase(unittest.TestCase):
+    def setUp(self):
+        self.place = (
+            paddle.CUDAPlace(0)
+            if fluid.is_compiled_with_cuda()
```
Comment (Aurelius84) [inline test/dygraph_to_static/test_pylayer.py]: 下面的fluid.xx都可以改成paddle.xxx或paddle.staitc.xxx

File: test/dygraph_to_static/test_pylayer.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+        )
+        self.to_static = False
+
+    def _run(self, *input_args, **input_kwargs):
+        assert getattr(
+            self, "dygraph_func", None
+        ), "Please setting `self.dygraph_func` before calling `self._run`"
+
+        paddle.jit.enable_to_static(self.to_static)
+        with fluid.dygraph.guard(self.place):
```
Comment (Aurelius84) [inline test/dygraph_to_static/test_pylayer.py]: 不需要使用 fluid.dygraph.guard了，切换device可以使用 paddle.set_device('gpu')这样

**Review by Aurelius84 — COMMENTED** [review]
Great work!

File: paddle/fluid/framework/executor.cc (hunk 1 above)
Comment (MarioLulab) [reply paddle/fluid/framework/executor.cc]: 确实不需要再对 excutor 进行适配了，之后会删除。目前使用的是 ProgramInterpreter，其会在 BuildOpFuncList 时调用 PrepareSafeEagerDeletionOnPyLayerOpAndPyLayerGradOp，所以只需要对目前使用到的 ProgramInterpreter 进行适配即可

File: paddle/fluid/framework/ir/graph_helper.cc (hunk 2 above)
Comment (MarioLulab) [reply paddle/fluid/framework/ir/graph_helper.cc]: 有道理，可以借助 unordered_set\<std::string\> 实现判断，提升代码可读性

File: paddle/fluid/framework/ir/memory_optimize_pass/pylayer_op_eager_deletion_pass.cc (hunk 3 above)
Comment (MarioLulab) [reply paddle/fluid/framework/ir/memory_optimize_pass/pylayer_op_eager_deletion_pass.cc]: 全局替换的时候没仔细检查 😢 之后会对其他部分做类似的检查

File: paddle/fluid/operators/controlflow/pylayer_op_helper.cc (hunk 4 above)
Comment (MarioLulab) [reply paddle/fluid/operators/controlflow/pylayer_op_helper.cc]: 是的，又是一个 typo ... 将修改

File: python/paddle/jit/dy2static/py_layer.py (hunk 6 above)
Comment (MarioLulab) [reply python/paddle/jit/dy2static/py_layer.py]: 有道理，之后将直接 return 布尔表达式

File: python/paddle/static/nn/static_pylayer.py (hunk 7 above)
Comment (MarioLulab) [reply python/paddle/static/nn/static_pylayer.py:15]: `copy_var_from_parent_block` 函数使用到了 `paddle.assign`. 我看到 control_flow.py 里也 `import paddle`，请问这样直接 `import paddle` 会有什么影响吗？

File: test/dygraph_to_static/test_pylayer.py (hunk 8 above)
Comment (MarioLulab) [reply test/dygraph_to_static/test_pylayer.py]: 已修改

File: test/dygraph_to_static/test_pylayer.py (hunk 9 above)
Comment (MarioLulab) [reply test/dygraph_to_static/test_pylayer.py]: 已修改

File: test/dygraph_to_static/test_pylayer.py (hunk 10 above)
Comment (MarioLulab) [reply test/dygraph_to_static/test_pylayer.py]: 已修改

File: python/paddle/static/nn/static_pylayer.py (hunk 7 above)
Comment (MarioLulab) [reply python/paddle/static/nn/static_pylayer.py:15]: paddle.assign 使用到了 paddle ，所以经讨论，此处的 import paddle 可保留

File: python/paddle/jit/dy2static/convert_operators.py (hunk 5 above)
Comment (MarioLulab) [reply python/paddle/jit/dy2static/convert_operators.py]: 有道理，这三个分支都是基于 in_declarative_mode() 判断当前是否处于动转静状态。下一个 commit 会进行修改

**Review by Aurelius84 — APPROVED** [review]
Great job!

**Review by sunzhongkai588 — APPROVED** [review]
LGTM～

**Review by zhangbo9674 — APPROVED** [review]
LGTM for interpreter_util

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**